crates/bevy_reflect/src/tuple_struct.rs

```diff
@@ -0,0 +1,212 @@
+        self.fields.get_mut(index).map(|field| &mut **field)
+    }
+
+    #[inline]
+    fn field_len(&self) -> usize {
+        self.fields.len()
+    }
+
+    #[inline]
+    fn iter_fields(&self) -> TupleStructFieldIter {
+        TupleStructFieldIter {
+            tuple_struct: self,
+            index: 0,
+        }
+    }
+
+    fn clone_dynamic(&self) -> DynamicTupleStruct {
+        DynamicTupleStruct {
+            name: self.name.clone(),
+            fields: self
+                .fields
+                .iter()
+                .map(|value| value.clone_value())
+                .collect(),
+        }
+    }
+}
+
+impl Reflect for DynamicTupleStruct {
+    #[inline]
+    fn type_name(&self) -> &str {
+        std::any::type_name::<Self>()
+    }
+
+    #[inline]
+    fn any(&self) -> &dyn Any {
+        self
+    }
+
+    #[inline]
+    fn any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    #[inline]
+    fn clone_value(&self) -> Box<dyn Reflect> {
+        Box::new(self.clone_dynamic())
+    }
+
+    #[inline]
+    fn reflect_ref(&self) -> ReflectRef {
+        ReflectRef::TupleStruct(self)
+    }
+
+    #[inline]
+    fn reflect_mut(&mut self) -> ReflectMut {
+        ReflectMut::TupleStruct(self)
+    }
+
+    fn apply(&mut self, value: &dyn Reflect) {
+        if let ReflectRef::TupleStruct(tuple_struct) = value.reflect_ref() {
+            for (i, value) in tuple_struct.iter_fields().enumerate() {
+                if let Some(v) = self.field_mut(i) {
+                    v.apply(value)
+                }
+            }
+        } else {
+            panic!("attempted to apply non-TupleStruct type to TupleStruct type");
+        }
+    }
+
+    fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>> {
+        *self = value.take()?;
+        Ok(())
+    }
+
+    fn hash(&self) -> Option<u64> {
+        None
+    }
+
+    fn partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
+        tuple_struct_partial_eq(self, value)
+    }
+
+    fn serializable(&self) -> Option<Serializable> {
+        None
+    }
+}
+
+#[inline]
+pub fn tuple_struct_partial_eq<S: TupleStruct>(a: &S, b: &dyn Reflect) -> Option<bool> {
+    Some(
+        if let ReflectRef::TupleStruct(tuple_struct) = b.reflect_ref() {
+            if a.field_len() == tuple_struct.field_len() {
+                for (i, value) in tuple_struct.iter_fields().enumerate() {
+                    if let Some(field_value) = a.field(i) {
+                        match field_value.partial_eq(value) {
+                            Some(true) => {}
+                            Some(false) => return Some(false),
+                            None => return Some(false),
+                        }
+                    } else {
+                        return Some(false);
+                    }
+                }
+
+                true
+            } else {
+                false
+            }
+        } else {
+            false
+        },
+    )
+}
```

Comment:
The control flow might be worth flattening a bit, something along the lines of the following. I also noticed that it `None` is ever returned. Should this return `bool` instead then be wrapped by `Some` outside of the function?
```suggestion
pub fn tuple_struct_partial_eq<S: TupleStruct>(a: &S, b: &dyn Reflect) -> Option<bool> {
    let tuple_struct = match b.reflect_ref() {
        Some(tuple_struct) => tuple_struct,
        None => return Some(false),
    };
    
    if a.field_len() != tuple_struct.field_len() {
        return Some(false);
    }
    
    for (i, value) in tuple_struct.iter_fields().enumerate() {
        let field_value = match a.field(i) {
            Some(field_value) => field_value,
            None => return Some(false),
        };
        if let Some(false) | None = field_value.partial_eq(value) {
            return Some(false);
        }
    }

    Some(true)
}
```

Reply:
I definitely dig the flatter structure. And yeah we could wrap it in Some for calling functions. My rationale for not doing that is that this was intended to be the "default" partial_eq reflect impl for tuple structs. Every caller is intended to be in that context, so why make them all wrap the result in the same way?

Reply:
I don't have strong feelings one way or the other for the return type. This feels like a matter of "aesthetic preference"